File: rss.rb
#!/usr/bin/ruby
# This script can update the title page from any RSS feed.
#
# It replaces the content between <!-- BLOG_HEADLINES_START -->
# and <!-- BLOG_HEADLINES_END --> in _site/index.html

require 'rss'

# Templating
require 'liquid'

# For striping HTML tags from post descriptions
require 'action_view'

# Following adjustments are based on Fedora Planet RSS feed.
# If you are changing the feed make sure the changes still apply!
FEED_URL = 'http://fedoraplanet.org/rss20.xml'.freeze

class Article
  attr_accessor :author, :title, :description, :date, :url
  liquid_methods :author, :title, :description, :date, :url

  def initialize(author, title, description, date, url)
    @author, @title, @description, @date, @url = author, title, description, date, url
  end
end
@articles = []

puts "Fetching #{FEED_URL} feed..."
rss = RSS::Parser.parse(FEED_URL, false)
rss.items.take(4).each do |item|
  # Extract name from the title
  #  Original title: NAME: TITLE
  #  Expected title: TITLE
  #  Expected name: NAME
  author = item.title.gsub(/([^:]*):(.*)/, '\\1').strip
  title = item.title.gsub(/([^:]*):(.*)/, '\\2').strip

  # Avoid HTML in description
  description = ActionView::Base.full_sanitizer.sanitize(item.description)

  # Shorter description if necessary
  if description && description.length > 140
    # Take whole words rather than xy letters
    description = description.split[0...25].join(' ')
    description += '&hellip;' unless ['!', '?', '.'].include? description[-1]
  end

  # Adjust date, strip time
  #   Original date: Tue, 08 Sep 2015 10:54:28 +0000
  #   Expected date: Tue, 08 Sep 2015
  regexp = /([a-zA-Z]{3}\, [0-9]{2} [a-zA-Z]{3} [0-9]{4}).*/
  date = item.pubDate.to_s.gsub(regexp, '\\1')

  @articles << Article.new(author, title, description, date, item.link)
end

template = <<TEMPLATE
<!-- BLOG_HEADLINES_START -->
<div class="container" id="blog-headlines">
<div class="container">
<div class="row">
  <div class="col-sm-12">
      <h2><span>Other headlines from our blog</span></h2>
  </div>
</div>
<div class="row">
  <div class="col-sm-6 blog-headlines">
  <article>
  <h3><a href="{{ articles[0].url }}">{{ articles[0].title }}</a></h3>
  <p>{{ articles[0].description }}</p>
  <p><a href="{{ articles[0].url }}">Read more&hellip;</a></p>
  <p class="byline">by <span class="author">{{ articles[0].author }}</span> <span class="date">{{ articles[0].date }}</span></p>
  </article>
  <article>
  <h3><a href="{{ articles[1].url }}">{{ articles[1].title }}</a></h3>
  <p>{{ articles[1].description }}</p>
  <p><a href="{{ articles[1].url }}">Read more&hellip;</a></p>
  <p class="byline">by <span class="author">{{ articles[1].author }}</span> <span class="date">{{ articles[1].date }}</span></p>
  </article>
  </div>
  <div class="col-sm-6 blog-headlines">
    <article>
  <h3><a href="{{ articles[1].url }}">{{ articles[2].title }}</a></h3>
  <p>{{ articles[2].description }}</p>
  <p><a href="{{ articles[2].url }}">Read more&hellip;</a></p>
  <p class="byline">by <span class="author">{{ articles[2].author }}</span> <span class="date">{{ articles[2].date }}</span></p>
  </article>
  <article>
  <h3><a href="{{ articles[3].url }}">{{ articles[3].title }}</a></h3>
  <p>{{ articles[3].description }}</p>
  <p><a href="{{ articles[3].url }}">Read more&hellip;</a></p>
  <p class="byline">by <span class="author">{{ articles[3].author }}</span> <span class="date">{{ articles[3].date }}</span></p>
  </article>
  </div>
</div>
</div>
</div>
<!-- BLOG_HEADLINES_END -->
TEMPLATE

blog_posts = Liquid::Template.parse(template).render 'articles' => @articles

INDEX_FILE = File.expand_path('_site/index.html', '.')
contents = File.open(INDEX_FILE).read
contents.gsub!(/<!-- BLOG_HEADLINES_START -->.*<!-- BLOG_HEADLINES_END -->/im, "\\1#{blog_posts}\\3")

File.open(File.expand_path(INDEX_FILE, '.'), 'w') { |file|
  file.write(contents)
}

File: rss.py
#!/usr/bin/python -tt
# -*- coding: utf-8 -*-

import codecs
import os
import sys
import feedparser
import re

defenc = "utf-8" if sys.getdefaultencoding() == "ascii" else sys.getdefaultencoding()

FedMag = ['http://fedoraplanet.org/rss20.xml']


HTML = u"""
"""


for feed in map(feedparser.parse, FedMag):
    # We will parse last ten items
    HTML += u"""
<div class="container" id="blog-headlines">
    <div class="container">
    <div class="row">
      <div class="col-sm-12">
          <h2><span>Other headlines from our blog</span></h2>
      </div>
    </div>
    <div class="row">
"""
    cnt = 0
    for item in feed["items"][:4]:
        if int(cnt) % 2 == 0:
            HTML += u"""
    <div class="col-sm-6 blog-headlines">
    """
        item.title = item.title.replace("&", "&#38;")
        author, title = item.title.split(':')
        link = item.links[0]['href']
        # Remove image tag from beginning
        article_desc = '\n'.join(item.description.split('\n')[1:])
        if len(article_desc) > 140:
            article_desc = ' '.join(article_desc.split()[0:25]) + '...'
        # we got
        # Tue, 20 Oct 2015 03:28:42 +0000
        # But we expect
        # Tue, 20 Oct 2015
        article_date = ' '.join(item.updated.split()[:4])
        HTML += u"""
        <article>
        <h3><a href="{article_url}">{article_title}</a></h3>
        <p>{article_desc}</p>
        <p><a href="{article_url}">Read more</a></p>
        <p class="byline">by <span class="author">{author}</span> <span class="date">{article_date}</span></p>
        </article>
""".format(article_url=link,
           article_title=title,
           article_desc=article_desc,
           article_date=article_date,
           author=author)
        cnt += 1
        if int(cnt) % 2 == 0:
            HTML += u"""
    </div>
"""
    HTML += u"""
</div>
</div>
</div>
"""

INDEX_FILE = os.path.join('.', '_site', 'index.html')
with codecs.open(INDEX_FILE, 'r', 'utf8') as f:
    contents = [line for line in f.readlines()]
if contents:
    with codecs.open(INDEX_FILE, 'w', 'utf8') as f:
        found_start = False
        for line in contents:
            if not found_start:
                f.write(line)
            if '<!-- BLOG_HEADLINES_START -->' in line:
                f.write(HTML)
                found_start = True
                continue
            if '<!-- BLOG_HEADLINES_END -->' in line:
                found_start = False
                continue

    #regexp = r'.*(<!-- BLOG_HEADLINES_START -->)(.*)(<!-- BLOG_HEADLINES_END -->).*'
    #print (re.search(regexp, contents, re.MULTILINE | re.DOTALL))
    #contents = re.sub(regexp, r'\1 MYREPLACE \3', contents, re.DOTALL | re.MULTILINE)
